pkg/oc/bootstrap/clusterup/components/router/router_install.go

```diff
@@ -0,0 +1,186 @@
+package router
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"k8s.io/client-go/util/retry"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+	kclientcmd "k8s.io/client-go/tools/clientcmd"
+	kapi "k8s.io/kubernetes/pkg/apis/core"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+
+	"github.com/openshift/origin/pkg/cmd/server/admin"
+	"github.com/openshift/origin/pkg/oc/bootstrap/clusterup/componentinstall"
+	"github.com/openshift/origin/pkg/oc/bootstrap/docker/dockerhelper"
+	"github.com/openshift/origin/pkg/oc/bootstrap/docker/run"
+	"github.com/openshift/origin/pkg/oc/errors"
+	securityclientinternal "github.com/openshift/origin/pkg/security/generated/internalclientset"
+)
+
+const (
+	DefaultNamespace         = "default"
+	RouterServiceAccountName = "router"
+	RouterServiceName        = "router"
+)
+
+type RouterComponentOptions struct {
+	OCImage         string
+	MasterConfigDir string
+	ImageFormat     string
+	PublicMasterURL string
+	RoutingSuffix   string
+	PortForwarding  bool
+}
+
+func (c *RouterComponentOptions) Name() string {
+	return "openshift-router"
+}
+
+func (c *RouterComponentOptions) Install(dockerClient dockerhelper.Interface, logdir string) error {
+	clusterAdminKubeConfigBytes, err := ioutil.ReadFile(path.Join(c.MasterConfigDir, "admin.kubeconfig"))
+	if err != nil {
+		return err
+	}
+	restConfig, err := kclientcmd.RESTConfigFromKubeConfig(clusterAdminKubeConfigBytes)
+	if err != nil {
+		return err
+	}
+	kubeClient, err := kclientset.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+	_, err = kubeClient.Core().Services(DefaultNamespace).Get(RouterServiceName, metav1.GetOptions{})
```

Comment:
Is this check enough? We do the same for registry I think. I was thinking that we always create all components, just ignore the 'AlreadyExists' errors. That matches what `oc create` does?

Reply:
> Is this check enough? We do the same for registry I think. I was thinking that we always create all components, just ignore the 'AlreadyExists' errors. That matches what oc create does?

Prexisting.  I don't plan to rewrite the oc adm router and registry commands.  I think this stays until they create operators.